server/project/TypescriptBuilder.scala

```diff
@@ -0,0 +1,114 @@
+import com.typesafe.sbt.web.Import.WebKeys.webTarget
+import com.typesafe.sbt.web.{PathMapping, incremental}
+import com.typesafe.sbt.web.pipeline.Pipeline
+import sbt.Keys.streams
+import com.typesafe.sbt.web.Import._
+import com.typesafe.sbt.web.incremental.{
+  OpInputHash,
+  OpInputHasher,
+  OpResult,
+  OpSuccess
+}
+import sbt.internal.util.ManagedLogger
+import sbt.{AutoPlugin, File, Setting, sbtUnchecked, taskKey}
+
+import scala.sys.process.{Process, ProcessLogger}
+
+/** Custom plugin to compile TypeScript code. We are not using sbt-typescript
+  * https://github.com/joost-de-vries/sbt-typescript plugin because it hasn't
+  * been updated since 2018 and uses Typescript 2 while current Typescript
+  * version is 4.
+  *
+  * Our custom plugin invokes `tsc` compiler directly. `tsconfig.json` is setup
+  * to compile code from `app/assets/javascripts` directory and outputs result
+  * to `target/web/typescript/javascripts` folder.
+  *
+  * This plugin introduces `compileTypescript` task which will be used as a
+  * pipeline stage. Read about Asset Pipeline tasks
+  * https://github.com/sbt/sbt-web#asset-pipeline The task implements Asset
+  * Pipeline task and not Source File task because the latter doesn't allow us
+  * to keep ∂irectory structure: we want that file `assets/javascripts/foo.ts`
+  * be compiled to `assets/javascripts/foo.js`. So we are using Pipeline.Stage
+  * for that.
+  *
+  * This plugin is enabled by adding `compileTypescript` as first task in
+  * pipelineStages in build.sbt.
+  */
+object TypescriptBuilder extends AutoPlugin {
+  override def trigger = allRequirements
+
+  object autoImport {
+    val compileTypescript = taskKey[Pipeline.Stage](
+      "Compiles typescript from assets/javascript folder"
+    )
+  }
+  import autoImport._
```

Comment:
suggestion (blocking): most people reading this are going to be pretty unfamiliar with sbt and scala, a little more commenting may be helpful. Maybe mention where the entrypoint is for code execution in the plugin, for example

Reply:
Added more comments. Agree that scala and sbt are pretty unfriendly to noobies like me due the amount of DSL they rely.